packages/state-transition/src/cache/epochCache.ts

```diff
@@ -748,9 +771,74 @@ export class EpochCache {
     return isAggregatorFromCommitteeLength(committee.length, slotSignature);
   }
 
+  /**
+   * Return finalized pubkey given the validator index.
+   * Only finalized pubkey as we do not store unfinalized pubkey because no where in the spec has a
+   * need to make such enquiry
+   *
+   */
+  getPubkey(index: ValidatorIndex): PublicKey {
+    return this.index2pubkey[index];
+  }
+
+  getValidatorIndex(pubkey: Uint8Array | PubkeyHex): ValidatorIndex | undefined {
+    if (this.isAfterEIP6110()) {
+      return this.pubkey2index.get(pubkey) ?? this.unfinalizedPubkey2index.get(toMemoryEfficientHexStr(pubkey));
+    } else {
+      return this.pubkey2index.get(pubkey);
+    }
+  }
+
+  /**
+   *
+   * Add unfinalized pubkeys
+   *
+   */
   addPubkey(index: ValidatorIndex, pubkey: Uint8Array): void {
+    if (this.isAfterEIP6110()) {
+      this.unfinalizedPubkey2index = this.unfinalizedPubkey2index.set(toMemoryEfficientHexStr(pubkey), index);
+    } else {
+      this.pubkey2index.set(pubkey, index);
+      this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
+    }
+  }
+
+  addFinalizedPubkeys(pubkeyMap: UnfinalizedPubkeyIndexMap): void {
+    Array.from(pubkeyMap.entries()).forEach(([pubkey, index]) => this.addFinalizedPubkey(index, pubkey));
+  }
+
+  /**
+   * Add finalized validator index and pubkey into finalized cache.
+   * Since addFinalizedPubkey() primarily takes pubkeys from unfinalized cache, it can take pubkey hex string directly
+   */
+  addFinalizedPubkey(index: ValidatorIndex, pubkey: PubkeyHex): void {
+    if (!this.isAfterEIP6110()) {
+      throw new Error("addFInalizedPubkey is not available pre EIP-6110");
+    }
+
+    const existingIndex = this.pubkey2index.get(pubkey);
+
+    if (existingIndex != undefined) {
+      if (existingIndex === index) {
+        // Repeated insert. Should not happen except during the first few epochs of 6110 activation
+        // Unfinalized validator added to finalizedPubkey2index pre-6110 by calling addPubkey()
+        // when it becomes finalized in post-6110, addFinalizedPubkey() is called to cause repeated insert
+        return;
```

Comment:
I think this condition will be hit many more times, since you are updating the global cache multiple times (once for each state). Can you add metrics to compute the total times this condition is hit?

Reply:
bump to add metrics